### PR TITLE
Store regex for attribute names in one place. Allow '+' character.

### PR DESCRIPTION
**Pull request by SamMousa**
Partial fix. Moves the regex to one place and adds support for the `+` character.
Does not add support for other non-word characters as mentioned in #4408.

| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #4408

### CODE REVIEW

**Comment by SamMousa** [comment]
I"m not sure the failed tests are caused by my code changes...

Can someone restart the tests? 
If the tests pass I think this is something that can be merged relatively quickly since it is only a few lines of code and will not break any existing projects.

It also allows for more advanced local configuration (by changing `BaseHtml::$attributeRegex`).

File: framework/helpers/BaseHtml.php [hunk 1]
```diff
@@ -24,6 +24,8 @@
  */
 class BaseHtml
 {
+
+    public static $attributeRegex = '/(^|.*\])([\w\.\+]+)(\[.*|$)/';
```
Comment (SilverFire) [inline framework/helpers/BaseHtml.php]: Could you add since tag and description?

**Review by SilverFire — CHANGES_REQUESTED** [review]

**Comment by kmindi** [comment]
Pls add a unit test as well, you can use or base yours on the one I proposed in the WIP PR https://github.com/yiisoft/yii2/pull/4446/files
Also I remember there was a comment by @cebe https://github.com/yiisoft/yii2/pull/4446#issuecomment-59438335 that states:
> attributes should not be exactly the same as allowed in gridview because gridview allows relational attribute names using . so ... must not be a valid name in Html.

**Comment by SamMousa** [comment]
Will do!

File: framework/helpers/BaseHtml.php [hunk 2]
```diff
@@ -24,6 +24,8 @@
  */
 class BaseHtml
 {
+
+    public static $attributeRegex = '/(^|.*\])([\w\.\+]+)(\[.*|$)/u';
```
Comment (samdark) [inline framework/helpers/BaseHtml.php]: Need phpdoc.

**Review by samdark — CHANGES_REQUESTED** [review]

**Comment by SamMousa** [comment]
@kmindi Have added the unit test, based on your test cases. Have added your name to the changelog as well.

One of your test cases contained a `,` which is a non-word character so I've added that to the error cases instead!

**Comment by SamMousa** [comment]
@samdark As usual these fixes are part of a feature we need in production ;-), any idea when 2.0.12 will be released?

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
https://github.com/yiisoft/yii2/milestones/2.0.12

File: tests/framework/helpers/HtmlTest.php [hunk 3]
```diff
@@ -1260,6 +1260,55 @@ public function testActiveCheckbox($value, array $options, $expectedHtml)
...
+
+    /**
+     * @dataProvider dataProviderAttributeNames
+     *
+     * @param string $name
+     * @param string|null $expected
+     */
+    public function testAttributeNameValidation($name, $expected)
+    {
+        if (!isset($expected)) {
```
Comment (dynasource) [inline tests/framework/helpers/HtmlTest.php]: it is not preferable to split logic in tests as tests should be instantly clear. Can you split this in 2 tests please?

**Review by dynasource — CHANGES_REQUESTED** [review]
good changes. Please split the unit tests towards best practice of 'single' purpose.

**Comment by SamMousa** [comment]
Will do

**Comment by SamMousa** [comment]
This is ready for merge.

File: tests/framework/helpers/HtmlTest.php [hunk 4]
```diff
@@ -1260,6 +1260,75 @@ public function testActiveCheckbox($value, array $options, $expectedHtml)
...
+     * Data provider for [[testAttributeNameValidation()]]
+     * @return array test data
+     */
+    public function dataProviderInvalidAttributeNames()
+    {
+        return [
+            '. ..',
+            'a +b',
+            'a,b',
+
```
Comment (dynasource) [inline tests/framework/helpers/HtmlTest.php]: blank line

File: tests/framework/helpers/HtmlTest.php [hunk 5]
```diff
@@ -1260,6 +1260,75 @@ public function testActiveCheckbox($value, array $options, $expectedHtml)
...
+            ["asdf]öáöio..[asdfasdf", "öáöio.."],
+            ["öáöio", "öáöio"],
+            ["[0]test.ööößß.d", "test.ööößß.d"],
+            ["ИІК", "ИІК"],
+            ["]ИІК[", "ИІК"],
+            ["[0]ИІК[0]", "ИІК"]
+        ];
+    }
+
+        /**
```
Comment (dynasource) [inline tests/framework/helpers/HtmlTest.php]: identation

File: tests/framework/helpers/HtmlTest.php [hunk 6]
```diff
@@ -1260,6 +1260,75 @@ public function testActiveCheckbox($value, array $options, $expectedHtml)
...
+            $this->assertEquals($expected, Html::getAttributeName($name));
+        }
+
+    }
+    
+    /**
+     * @dataProvider dataProviderInvalidAttributeNames
+     *
+     * @param string $name
+     * @param string|null $expected
```
Comment (dynasource) [inline tests/framework/helpers/HtmlTest.php]: @param is obsolete now

**Review by dynasource — CHANGES_REQUESTED** [review]
polishing required.

**Review by dynasource — APPROVED** [review]

**Comment by samdark** [comment]
https://travis-ci.org/yiisoft/yii2/jobs/211271201#L565

**Comment by SamMousa** [comment]
Currently it fails due to some database errors. Could travis be reconfigured to scan the output and retry if those are the errors?

**Comment by samdark** [comment]
Tests are fine. These DB errors are cryptic and hard to fix. It is retrying 3 times now but that doesn't really help.

**Comment by SamMousa** [comment]
Ah, I'm seeing the same on gitlab-ci builds :| 
Can travis be controlled via github comments? like @travis-ci retest please or something like that?

**Comment by samdark** [comment]
Merged. Thank you!

**Comment by samdark** [comment]
@SamMousa I can restart travis manually from project page.

**Pull request merged**